### PR TITLE
update the docs to match the latest thinking

### PR DESCRIPTION
**Pull request by bensigelman**
(per https://github.com/opentracing/opentracing.github.io/issues/44, https://github.com/opentracing/opentracing-go/pull/48, and https://github.com/opentracing/opentracing-go/pull/50)

### CODE REVIEW

File: spec.md [hunk 1]
```diff
@@ -49,36 +49,35 @@ Every Span has zero or more **Logs**, each of which being a timestamped event na
...
+Trace Attributes come with powerful _costs_ as well; since the attributes are propagated in-band, if they are too large or too numerous they may decrease system throughput or increase RPC latencies.
 
 **Trace Attributes** vs. **Span Tags**
 
-* Trace Attributes are propagated in-band (i.e., alongside the actual application data) across process boundaries. Span Tags are not propagated since they are not inherited from parent Span to child Span.
-* Span Tags are recorded out-of-band from the application data, presumably in the tracing system's storage. Trace Attributes are not necessarily recorded here, though an implementation may elect to.
+- Trace Attributes are propagated in-band (i.e., alongside the actual application data) across process boundaries. Span Tags are not propagated since they are not inherited from parent Span to child Span.
+- Span Tags are recorded out-of-band from the application data, presumably in the tracing system's storage. Trace Attributes are not necessarily recorded here, though an implementation may elect to.
```
Comment (yurishkuro) [inline spec.md]: "recorded _t_here" ?

File: spec.md [hunk 2]
```diff
@@ -92,39 +91,51 @@ The `Span` interface must have the following capabilities:
...
+- Return an `Injector` instance given a format identifier (see [Injector and Extractor](#injector-and-extractor))
+- Return an `Extractor` instance given a format identifie (see [Injector and Extractor](#injector-and-extractor))
 
-### SpanPropagator
+`Injector` and `Extractor` both rely on "format identifiers." The nature of the "format identifier" may vary from platform to platform, but in all cases the format identifiers should come from a global namespace. New formats must not *require* changes to the core OpenTracing platform APIs, though those core platform APIs should define a few basic/general formats (like string maps, binary blobs, and/or HTTP headers). For example, if the maintainer of EsotericRPCFramework wanted to define an EsotericRPCFramework injection and extraction format, she or he must be able to do so without sending a PR to OpenTracing maintainers (though of course OpenTracing implementations are not required to suppor the EsotericRPCFramework format).
```
Comment (yurishkuro) [inline spec.md]: move to Injector and Extractor section?

File: spec.md [hunk 3]
```diff
@@ -92,39 +91,51 @@ The `Span` interface must have the following capabilities:
...
+To make all the above more concrete, consider the following sequence:
+
+1. A client process has a `Span` instance and is about to make an RPC over a home-grown HTTP protocol
+1. That client process calls some `Tracer.Injector(...)` method, passing a format identifier for a string-to-string map
+1. With an `Injector` instance in hand, the client calls `injector.InjectSpan(...)`, passing the active `Span` instance and a string-to-string map carrier in as parameters
+1. The newly-populateed (opaque) string-to-string map is encoded into the homegrown HTTP protocol (e.g., as headers) by the application code
+1. Now in the server process, the application code grabs the opaque string-to-string map from the homegrown HTTP protocol
+1. The server process calls some `Tracer.Extractor(...)` method, again passing a format identifier for a string-to-string map
+1. With an `Extractor` instance in hand, the server calls `extractor.JoinTrace(...)`, passing in the desired operation name and string-to-string map carrier as parameters
+1. In the absence of data corruption or other errors, the server now has a `Span` instance that belongs to the same trace as the one in the client
```
Comment (yurishkuro) [inline spec.md]: I would refer to use-cases.md here for concrete code examples.

File: use-cases.md [hunk 4]
```diff
@@ -170,26 +170,21 @@ def traced_request(request, operation, http_client):
...
-    if parent_span is None:
-        span = tracer.start_trace(operation_name=operation)
-    else:
-        span = parent_span.start_child(operation_name=operation)
-    span.set_tag('http.url', request.full_url)
+    # start a new span to represent the RPC
+    span = tracer.start_span(
+        operation_name=operation,
+        parent=parent_span,
+        tags={'http.url': request.full_url
```
Comment (yurishkuro) [inline use-cases.md]: missing closing `}`

File: spec.md [hunk 5]
```diff
@@ -92,39 +91,51 @@ The `Span` interface must have the following capabilities:
...
+- An `Injector` has one method: `InjectSpan`. InjectSpan takes two arguments: a `Span` instance and an "carrier" object in which to inject that `Span` for cross-process propagation. The type of the carrier depends on the particular type of `Injector`, and the type of the `Injector` in turn depends on the "format identifier".
 
-The encoded data is separated in this way for a variety of reasons; the most important is to give OpenTracing users a portable way to opt out of Trace Attribute propagation entirely if they deem it a stability risk.
+- An `Extractor` has one method: `JoinTrace`. JoinTrace takes two arguments: the operation name for the `Span` it's about to create and a format-specific "carrier" object from which to extract identifying information needed by the new `Span` instance. Unless there's an error, it returns a freshly-started `Span` which can be used in the host process like any other. (Note that some OpenTracing implementations consider the `Span`s on either side of an RPC to have the same identity, and others consider the caller to be the parent and the receiever to be the child)
```
Comment (bg451) [inline spec.md]: s/receiever/receiver

File: use-cases.md [hunk 6]
```diff
@@ -170,26 +170,21 @@ def traced_request(request, operation, http_client):
...
-    h_ctx, h_attr = tracer.propagate_span_as_text(span)
-
-    for key, value in h_ctx.iteritems():
-        request.add_header(key, value)
-    if h_attr:
-        for key, value in h_attr.iteritems():
-            request.add_header(key, value)
+    # propagate the Span via HTTP request headers
+    injector = tracer.injector(opentracing.HTTP_HEADER_FORMAT)
+    injector.inject(span, request)
```
Comment (yurishkuro) [inline use-cases.md]: request.Header ?

File: use-cases.md [hunk 7]
```diff
@@ -235,17 +229,28 @@ The tracer automatically records a timestamp of the event, in contrast with tags
...
+explicit_span.finish(
+    finish_time=external_format.finish,
+    bulk_logs=map(..., external_format.logs)
+)
+{% endhighlight %}
 
 ### Setting "Debug" Mode
 
-`TODO: re-introduce an API for this when we tackle explicit Span timestamps.`
+Most tracing systems apply sampling to minimize the amount of trace data sent to the system.  Sometimes developers want to have a way to ensure that a particular trace is going to be recorded (sampled) by the tracing system, e.g. by including a special parameter in the HTTP request, like `debug=true`. The OpenTracing API does not have any insight into sampling techniques used by the implementation, so there is no explicit API to force it. However, the implementations are advised to recognized the `debug` Trace Attribute and take measures to record the Span. In order to pass this attribute to tracing systems that rely on pre-trace sampling, the following approach can be used:
```
Comment (yurishkuro) [inline use-cases.md]: I think we converged to `sampling.priority` tag, not `debug` trace attribute.

**Comment by yurishkuro** [comment]
minor comments, otherwise LGTM

File: spec.md (hunk 2 above)
Comment (bensigelman) [inline spec.md]: I think I'll leave this here... it's really a comment about the "format identifiers" more than the Injectors/Extractors themselves and thus belongs in the Tracer section. I will reword, though.

File: spec.md (hunk 3 above)
Comment (bensigelman) [inline spec.md]: good plan.

File: use-cases.md (hunk 4 above)
Comment (bensigelman) [inline use-cases.md]: good catch!

File: spec.md (hunk 5 above)
Comment (bensigelman) [inline spec.md]: eagle :eyes:

File: use-cases.md (hunk 6 above)
Comment (bensigelman) [inline use-cases.md]: sure, done. added a `carrier=`, too.

File: use-cases.md (hunk 7 above)
Comment (bensigelman) [inline use-cases.md]: ah right -- fixed.

File: spec.md [hunk 8]
```diff
@@ -49,36 +49,35 @@ Every Span has zero or more **Logs**, each of which being a timestamped event na
 
 Every Span may also have zero or more key/value **Tags**, which do not have timestamps and simply annotate the spans.
 
-Spans may be **propagated**; that is, they may cross process boundaries along with sufficient information to continue the Trace on the far side of the propagation.
+Spans may be **Injected** into and **Extracted** from objects that are used for inter-process communication (e.g., HTTP headers). In this way, Spans may propagate across process boundaries along with sufficient information to join up with the Trace in some remote process.
```
Comment (adriancole) [inline spec.md]: This is very intuitive (to me). Nice job.

File: spec.md [hunk 9]
```diff
@@ -49,36 +49,35 @@ Every Span has zero or more **Logs**, each of which being a timestamped event na
...
+Trace Attributes come with powerful _costs_ as well; since the attributes are propagated in-band, if they are too large or too numerous they may decrease system throughput or increase RPC latencies.
 
 **Trace Attributes** vs. **Span Tags**
 
-* Trace Attributes are propagated in-band (i.e., alongside the actual application data) across process boundaries. Span Tags are not propagated since they are not inherited from parent Span to child Span.
-* Span Tags are recorded out-of-band from the application data, presumably in the tracing system's storage. Trace Attributes are not necessarily recorded here, though an implementation may elect to.
+- Trace Attributes are propagated in-band (i.e., alongside the actual application data) across process boundaries. Span Tags are not propagated since they are not inherited from parent Span to child Span.
+- Span Tags are recorded out-of-band from the application data, presumably in the tracing system's storage. Trace Attributes are not necessarily recorded there, though an implementation may elect to.
```
Comment (adriancole) [inline spec.md]: How about "implementations may choose to also record trace attributes"

File: spec.md [hunk 10]
```diff
@@ -49,36 +49,35 @@ Every Span has zero or more **Logs**, each of which being a timestamped event na
...
 
 
-## Platform-Independent API Semantics
+# Platform-Independent API Semantics
 
-OpenTracing supports a number of different platforms, and of course the per-platform APIs try to blend in to their surroundings and do as the Romans do. That said, each platform API must model a common set of semantics for the core tracing concepts described above. In this document we attempt to describe those concepts and semantics in a language- and platform-agnostic fashion.
+OpenTracing supports a number of different platforms, and of course the per-platform APIs try to blend into their surroundings and "do as the Romans do." That said, each platform API must model a common set of semantics for the core tracing concepts described above. In this document we attempt to describe those concepts and semantics in a language- and platform-agnostic fashion.
```
Comment (adriancole) [inline spec.md]: Late note, but non-english might struggle to unravel blending and doing as Romans bit. 

Maybe be more explicit: OT apis try to adhere to idioms and conventions of their language and platforms.

Feel free to punt if distracting

File: spec.md (hunk 9 above)
Comment (bensigelman) [inline spec.md]: done!

File: spec.md (hunk 10 above)
Comment (bensigelman) [inline spec.md]: good point, changed. :cn: :de: :fr: :ru: (etc)

File: spec.md [hunk 11]
```diff
@@ -92,39 +91,52 @@ The `Span` interface must have the following capabilities:
...
+## Injector and Extractor
 
-The encoded form of a propagated span is divided into two components:
+The `Injector` and `Extractor` interfaces allow Traces to propagate across process boundaries. They are not entirely symmetrical interfaces since `Span` injection is a fundamentally lossy process (most of the `Span`'s state is meant to be recorded out-of-band from the application's own communication path, whereas Inject-ed state is meant to be sent in-band with the application data).
 
-1. The core identifying information for the Span, referred to as the "context snapshot" (for example, in Dapper this would include a `trace_id`, a `span_id`, and a bitmask representing the sampling status for the given trace)
-1. Any trace attributes (per Span's ability to set trace attributes that propagate across process boundaries)
+- An `Injector` has one method: `InjectSpan`. InjectSpan takes two arguments: a `Span` instance and an "carrier" object in which to inject that `Span` for cross-process propagation. The type of the carrier depends on the particular type of `Injector`, and the type of the `Injector` in turn depends on the "format identifier".
```
Comment (adriancole) [inline spec.md]: @marcingrzejszczak fyi this is a recent change to OT, which makes it more clear where the propagation responsibility lies. Ex the Injected state, as listed above, is only the state send over headers or similar.

File: spec.md [hunk 12]
```diff
@@ -92,39 +91,52 @@ The `Span` interface must have the following capabilities:
...
 
-### Tracer
+## Tracer
 
 The `Tracer` interface must have the following capabilities:
 
-- Start a `Span` that has no parent, commonly referred to as a *root* `Span` **(py: `start_trace`, go: `StartTrace`)**
-- Provide some form of access to a SpanPropagator (whether that's through embedding, inheritance, containment, or something else is per-platform)
+- Start a new `Span`. There must be a way for the caller to specify a parent `Span`, an explicit start timestamp (other than "now"), and an initial set of `Span` tags. **(py: `start_span`, go: `StartSpanWithOptions`)**
+- Return an `Injector` instance given a format identifier, or nil/null if the format is not supported by the implementation (see [Injector and Extractor](#injector-and-extractor))
```
Comment (adriancole) [inline spec.md]: thinking out loud.. "format identifier" isn't intuitive, and it takes some navigation to find what it is.. maybe in this case, use an example?  Ex. "Return an `Injector` instance given a format identifier, like ???, or nill..."

File: spec.md [hunk 13]
```diff
@@ -92,39 +91,52 @@ The `Span` interface must have the following capabilities:
...
+- Return an `Extractor` instance given a format identifier, or nil/null if the format is not supported by the implementation (see [Injector and Extractor](#injector-and-extractor))
 
-### SpanPropagator
+A note about **"format identifiers"** per the `Injector` and `Extractor` lookup methods above: the nature of the "format identifier" may vary from platform to platform, but in all cases they should be drawn from a global namespace. New formats must not *require* changes to the core OpenTracing platform APIs, though those core platform APIs must define a few basic/general formats (like string maps and binary blobs). For example, if the maintainer of EsotericRPCFramework wanted to define an EsotericRPCFramework injection and extraction format, she or he must be able to do so without sending a PR to OpenTracing maintainers (though of course OpenTracing implementations are not required to suppor the EsotericRPCFramework format).
```
Comment (adriancole) [inline spec.md]: This really needs examples. for example nested in here is "few basic/general formats (like string maps and binary blobs)", but we don't give an example of what a format identifier might look like. pseudocode might help clear this up.

File: spec.md [hunk 14]
```diff
@@ -92,39 +91,52 @@ The `Span` interface must have the following capabilities:
...
 
-The encoded data is separated in this way for a variety of reasons; the most important is to give OpenTracing users a portable way to opt out of Trace Attribute propagation entirely if they deem it a stability risk.
+#### Required Injection/Extraction formats
 
-The propagation and trace-joining methods come in two flavors: binary and text.
+At a minimum, all platforms are expected to require implementations to support two formats: the "split text" format and the "split binary" format. The "split" refers to two common components of propagated spans:
+
+1. The core identifying information for the `Span`, referred to as the "tracer state" (for example, in Dapper this would include a `trace_id`, a `span_id`, and a bitmask representing the sampling status for the given trace)
```
Comment (adriancole) [inline spec.md]: love the example

File: spec.md [hunk 15]
```diff
@@ -92,39 +91,52 @@ The `Span` interface must have the following capabilities:
...
+Note that there is no expectation that different tracing systems Inject and Extract `Spans` in compatible ways. Though OpenTracing is agnostic about the tracing implementation, for successful inter-process handoff it's essential that the processes on both sides of a propagation use the same tracing implementation.
+
+#### An end-to-end propagation example
+
+To make all the above more concrete, consider the following sequence:
+
+1. A client process has a `Span` instance and is about to make an RPC over a home-grown HTTP protocol
+1. That client process calls some `Tracer.Injector(...)` method, passing a format identifier for a string-to-string map
+1. With an `Injector` instance in hand, the client calls `injector.InjectSpan(...)`, passing the active `Span` instance and a string-to-string map carrier in as parameters
+1. The newly-populateed (opaque) string-to-string map is encoded into the homegrown HTTP protocol (e.g., as headers) by the application code
```
Comment (adriancole) [inline spec.md]: populateed

File: spec.md [hunk 16]
```diff
@@ -92,39 +91,52 @@ The `Span` interface must have the following capabilities:
...
+1. Now in the server process, the application code grabs the opaque string-to-string map from the homegrown HTTP protocol
+1. The server process calls some `Tracer.Extractor(...)` method, again passing a format identifier for a string-to-string map
+1. With an `Extractor` instance in hand, the server calls `extractor.JoinTrace(...)`, passing in the desired operation name and string-to-string map carrier as parameters
```
Comment (adriancole) [inline spec.md]: "string-to-string map carrier" sounds quite jargony.. if this is a forged example about http, would there be a way to say something more concrete?

**Comment by adriancole** [comment]
LG.. just a few minor notes

**Comment by bensigelman** [comment]
@adriancole ec02769 -- thanks!

**Pull request merged**